packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3010,10 +3049,39 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         type: ContainerMessageType,
         contents: any,
         batch: boolean,
+        contentLength: number,
         appData?: any,
     ) {
         this.verifyNotClosed();
         assert(this.connected, 0x259 /* "Container disconnected when trying to submit system message" */);
+
+        if (contentLength > this.runtimeOptions.compressionOptions?.minimumSize) {
+            const originalContentLength = IsoBuffer.from(JSON.stringify(contents)).toString("base64").length;
+
+            const compressionStart = Date.now();
+            const stringifiedContent = JSON.stringify(contents);
+            const contentsAsBuffer = new TextEncoder().encode(stringifiedContent);
+            const compressedContents = compress(contentsAsBuffer);
+            const compressionTime = Date.now() - compressionStart;
+            const compressedContent = IsoBuffer.from(compressedContents).toString("base64");
```

Comment:
Note that we already have stringified content as part of chunking logic. Can we pass it around not to require strigification again? I.e. push stringification up the stack to reduce duplication on main path? I'd also include base64 encoding into time measurement.